test/testfloat/testfloat.cpp

```diff
@@ -0,0 +1,395 @@
+        {"f64_trunc_to_ui64", {31, Type::i64, Type::f64, Options::TrapIsInvalidOperation}},
+
+        {"f32_to_f64", {32, Type::f64, Type::f32}},
+        {"f64_to_f32", {33, Type::f32, Type::f64}},
+
+        {"f32_ceil", {34, Type::f32, Type::f32}},
+        {"f32_floor", {35, Type::f32, Type::f32}},
+        {"f32_trunc", {36, Type::f32, Type::f32}},
+        {"f32_nearest", {37, Type::f32, Type::f32}},
+        {"f64_ceil", {38, Type::f64, Type::f64}},
+        {"f64_floor", {39, Type::f64, Type::f64}},
+        {"f64_trunc", {40, Type::f64, Type::f64}},
+        {"f64_nearest", {41, Type::f64, Type::f64}},
+    };
+
+    if (const auto it = map.find(name); it != map.end())
+        return it->second;
+    throw std::invalid_argument{"unknown <function>: " + std::string{name}};
+}
+
+constexpr fizzy::Value make_value(Type type, uint64_t bits) noexcept
+{
+    switch (type)
+    {
+    case Type::i32:
+        return fizzy::Value{static_cast<uint32_t>(bits)};
+    case Type::i64:
+        return fizzy::Value{bits};
+    case Type::f32:
+        return fizzy::Value{FP{static_cast<uint32_t>(bits)}.value};
+    case Type::f64:
+        return fizzy::Value{FP{bits}.value};
+    }
+    __builtin_unreachable();
+}
+
+class TypedValue
+{
+    Type type_;
+    fizzy::Value value_;
+
+public:
+    constexpr TypedValue(Type type, fizzy::Value value) noexcept : type_{type}, value_{value} {}
+
+    constexpr TypedValue(Type type, uint64_t bits) noexcept
+      : type_{type}, value_{make_value(type, bits)}
+    {}
+
+    bool operator!=(uint64_t expected_bits) const
+    {
+        switch (type_)
+        {
+        case Type::i32:
+        case Type::i64:
+            return value_.i64 != expected_bits;
+        case Type::f32:
+            return value_.f32 != FP{static_cast<uint32_t>(expected_bits)};
+        case Type::f64:
+            return value_.f64 != FP{expected_bits};
+        }
+        __builtin_unreachable();
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const TypedValue& x)
+    {
+        os << std::hex << std::uppercase << std::setfill('0');
+        switch (x.type_)
+        {
+        case Type::i32:
+            return os << std::setw(8) << x.value_.as<uint32_t>();
+        case Type::i64:
+            return os << std::setw(16) << x.value_.i64;
+        case Type::f32:
+            return os << std::setw(8) << FP{x.value_.f32}.as_uint();
+        case Type::f64:
+            return os << std::setw(16) << FP{x.value_.f64}.as_uint();
+        }
+        __builtin_unreachable();
+    }
+};
+
+bool check(const FunctionDescription& func, fizzy::Instance& instance, const uint64_t inputs[],
+    bool ignore_nans)
+{
+    const auto report_failure = [&](auto result, auto expected) {
+        std::cerr << "FAILURE: " << result << " <-";
+        for (size_t i = 0; i < func.num_arguments; ++i)
+            std::cerr << ' ' << TypedValue{func.param_types[i], inputs[i]};
+        std::cerr << "\n   exp.: " << expected << "\n";
```

Comment:
I'd use `expected:` here.

Reply:
This looks like that:
```
FAILURE: 00000000 <- 00000000 00000000
   exp.: 00000001
```
Not the best in the world, but does its job without need of computing offsets.

Reply:
Not that much important, but `exp.` to me looks like exponent, though I have been working too much EVM. How about something like:
```
FAILURE:   0000
 expected: 0000
```

Reply:
```
FAILURE: 00000000 <- 00000000 00000000
         00000001 (expected)
```
This better?

Reply:
Looks good.

Reply:
Done.